src/Internal/Contract/QueryBackend.purs

```diff
@@ -0,0 +1,69 @@
+module Ctl.Internal.Contract.QueryBackend
+  ( BlockfrostBackend
+  , BlockfrostBackendParams
+  , CtlBackend
+  , CtlBackendParams
+  , QueryBackend(BlockfrostBackend, CtlBackend)
+  , QueryBackendParams(BlockfrostBackendParams, CtlBackendParams)
+  , getBlockfrostBackend
+  , getCtlBackend
+  , mkBlockfrostBackendParams
+  , mkCtlBackendParams
+  ) where
+
+import Prelude
+
+import Ctl.Internal.QueryM (OgmiosWebSocket)
+import Ctl.Internal.QueryM.ServerConfig (ServerConfig)
+import Data.Maybe (Maybe(Just, Nothing))
+
+--------------------------------------------------------------------------------
+-- QueryBackend
+--------------------------------------------------------------------------------
+
+data QueryBackend
+  = CtlBackend CtlBackend (Maybe BlockfrostBackend)
+  | BlockfrostBackend BlockfrostBackend (Maybe CtlBackend)
+
+type CtlBackend =
+  { ogmios ::
+      { config :: ServerConfig
+      , ws :: OgmiosWebSocket
+      }
+  , kupoConfig :: ServerConfig
+  }
+
+type BlockfrostBackend =
+  { blockfrostConfig :: ServerConfig
+  }
+
+getCtlBackend :: QueryBackend -> Maybe CtlBackend
+getCtlBackend (CtlBackend backend _) = Just backend
+getCtlBackend (BlockfrostBackend _ backend) = backend
+
+getBlockfrostBackend :: QueryBackend -> Maybe BlockfrostBackend
+getBlockfrostBackend (CtlBackend _ backend) = backend
+getBlockfrostBackend (BlockfrostBackend backend _) = Just backend
+
+--------------------------------------------------------------------------------
+-- QueryBackendParams
+--------------------------------------------------------------------------------
+
+data QueryBackendParams
+  = CtlBackendParams CtlBackendParams (Maybe BlockfrostBackendParams)
```

Comment:
Why do we need this `Maybe` param?

Reply:
The idea is to allow for the possibility of a "backup" backend, for any operations the main backend can't perform

Reply:
But PR does not actually use it? Is there some plans to use it in the future?